### PR TITLE
Add PackageVersion completion

### PR DESCRIPTION
**Pull request by RudySchockaert-EngieIT**
As of .Net 7 you can opt for [Central Package Management](https://learn.microsoft.com/en-us/nuget/consume-packages/Central-Package-Management).
A PackageReference can not have a Version attribute anymore, but the version must be specified on the PackageVersion item.
For autocompletion, the behavior for both PackageReference & PackageVersion are the same.

### CODE REVIEW

**Comment by tintoy** [comment]
Thanks for contributing! I’ll take a proper look at this later today 🙂

File: src/LanguageServer.Engine/CompletionProviders/ItemMetadataCompletionProvider.cs [hunk 1]
```diff
@@ -221,7 +221,7 @@ IEnumerable<CompletionItem> GetElementCompletions(XmlLocation location, ProjectD
             HandleTriggerCharacters(triggerCharacters, projectDocument, ref targetRange);
 
             // These items are handled by PackageReferenceCompletion.
-            if (itemType == "PackageReference" || itemType == "DotNetCliToolReference")
+            if (itemType == "PackageReference" || itemType == "PackageVersion" || itemType == "DotNetCliToolReference")
```
Comment (DoctorKrolic) [inline src/LanguageServer.Engine/CompletionProviders/ItemMetadataCompletionProvider.cs:224]:
```suggestion
            if (itemType is "PackageReference" or "PackageVersion" or "DotNetCliToolReference")
```

File: src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs [hunk 2]
```diff
@@ -173,7 +173,7 @@ public MarkedStringContainer ItemGroup(MSBuildItemGroup itemGroup)
             if (itemGroup == null)
                 throw new ArgumentNullException(nameof(itemGroup));
 
-            if (itemGroup.Name == "PackageReference")
+            if (itemGroup.Name == "PackageReference" || itemGroup.Name == "PackageVersion")
```
Comment (DoctorKrolic) [inline src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs:176]:
```suggestion
            if (itemGroup.Name is "PackageReference" or "PackageVersion")
```

File: src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs [hunk 3]
```diff
@@ -363,7 +363,7 @@ public MarkedStringContainer ItemGroupMetadata(MSBuildItemGroup itemGroup, strin
             if (string.IsNullOrWhiteSpace(metadataName))
                 throw new ArgumentException("Argument cannot be null, empty, or entirely composed of whitespace: 'metadataName'.", nameof(metadataName));
 
-            if (itemGroup.Name == "PackageReference")
+            if (itemGroup.Name == "PackageReference" || itemGroup.Name == "PackageVersion")
```
Comment (DoctorKrolic) [inline src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs:366]:
```suggestion
            if (itemGroup.Name is "PackageReference" or "PackageVersion")
```

File: src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs [hunk 4]
```diff
@@ -437,7 +437,7 @@ public MarkedStringContainer UnusedItemGroupMetadata(MSBuildUnusedItemGroup item
...
```
Comment (DoctorKrolic) [inline src/LanguageServer.Engine/ContentProviders/HoverContentProvider.cs:440]:
```suggestion
            if (itemGroup.Name is "PackageReference" or "PackageVersion")
```

File: src/LanguageServer.Engine/CompletionProviders/ItemMetadataCompletionProvider.cs (hunk 1 above)
Comment (tintoy) [reply src/LanguageServer.Engine/CompletionProviders/ItemMetadataCompletionProvider.cs:224]: Isn't the `is` operator normally for type-testing? I get that you can use it with the new pattern matching features but is it really clearer than `==`?

**Comment by tintoy** [comment]
Thanks! This enhancement looks good to me, but maybe consider only offering the `PackageVersion` element completion if the MSBuild project's `ManagePackageVersionsCentrally` property is set to `true`? Otherwise people may select it and then wonder why it has no effect.

You might also want to consider renaming the `HandlePackageReferenceElementCompletion` and `HandlePackageReferenceAttributeCompletion` methods to `HandleElementCompletion` and `HandleAttributeCompletion` since the handler now deals with multiple item types.

**Comment by RudySchockaert-EngieIT** [comment]
I like this proposal.
It is easier to read indeed, but just as Adam mentioned already the ‘is’ operator had a Type feeling for me too.

I’m now working on the test setup inluding the vscode extension before I’ll adapt the PR.

--
SCHOCKAERT Rudy
Expertise User Solutions
Digital Workplace
GBS IT

***@***.******@***.***>
T : + 32 2 206 39 18
M: + 32 473 84 69 08
_________________________


***@***.***

engie.com

<http://www.engie.com/>
36, Boulevard Simon Bolivar (ETB2)
Office G10BC

1000 Bruxelles - Belgique

**Comment by RudySchockaert-EngieIT** [comment]
I added the change as propoised by [@DoctorKrolic](https://github.com/DoctorKrolic)

**Comment by tintoy** [comment]
Thanks! I’ll take a look at this first thing tomorrow morning 🙂

**Comment by tintoy** [comment]
This looks good to me; I'm happy to merge it and then make any [additional changes](https://github.com/tintoy/msbuild-project-tools-server/pull/84#issuecomment-1836933273) separately, if needed. @DoctorKrolic sound good to you?

**Review by DoctorKrolic — APPROVED** [review]
Ok, LGTM

**Comment by tintoy** [comment]
Thanks, @RudySchockaert-EngieIT !

**Pull request merged**